base/REPL.jl

```diff
@@ -37,32 +37,44 @@ abstract AbstractREPL
 answer_color(::AbstractREPL) = ""
 
 type REPLBackend
-    repl_channel::Channel
-    response_channel::Channel
-    in_eval::Bool
-    ans
-    backend_task::Task
-    REPLBackend(repl_channel, response_channel, in_eval, ans) =
-        new(repl_channel, response_channel, in_eval, ans)
+    repl_channel::Channel # channel for AST
+    response_channel::Channel # channel for results: (value, nothing) or (error, backtrace)
+    task::Task # current backend task
+    responded::Bool # flag indicating the state of this backend
 end
 
+function put_backend_response(backend::REPLBackend, response, validsource::Bool)
+    if validsource && !backend.responded
+        backend.responded = true
+        put!(backend.response_channel, response)
+    else
+        @schedule begin
+            println(STDERR, "Got a REPL eval result, but already returned a result:")
+            value, bt = response
+            if bt !== nothing
+                showerror(STDERR, value, bt)
+            else
+                show(STDERR, value)
+            end
+        end
+    end
+    nothing
+end
+
+# note: the name of this function is significant for printing error backtraces
 function eval_user_input(ast::ANY, backend::REPLBackend)
-    iserr, lasterr, bt = false, (), nothing
-    while true
+    iserr, lasterr = false, (nothing, ())
+    while !backend.responded
         try
             if iserr
-                put!(backend.response_channel, (lasterr, bt))
+                put_backend_response(backend, lasterr, backend.task == current_task())
                 iserr, lasterr = false, ()
             else
-                ans = backend.ans
                 # note: value wrapped in a non-syntax value to avoid evaluating
                 # possibly-invalid syntax (issue #6763).
-                eval(Main, :(ans = $(getindex)($(Any[ans]), 1)))
-                backend.in_eval = true
                 value = eval(Main, ast)
-                backend.in_eval = false
-                backend.ans = value
-                put!(backend.response_channel, (value, nothing))
+                eval(Main, :(ans = $(Expr(:quote, value))))
```

Comment:
i hadn't, but trying it now it seems to work as expected. it looks like i need to delete the above comment too.

i was modifying this on the implementation of the similar function in client.jl:
https://github.com/JuliaLang/julia/commit/07a0c099596385d3cca0ceb3b9fffadf84cca2b3